File: app/src/main/java/com/github/pokatomnik/frontendconf2023/examples/ex09disposableeffects/ex09DisposableEffects.kt
package com.github.pokatomnik.frontendconf2023.examples.ex09disposableeffects

import androidx.compose.runtime.Composable
import androidx.compose.runtime.DisposableEffect

fun toastStub(message: String) {
    // Показываем здесь сообщение
}

interface PubSub {
    fun subscribe(
        subscriber: (value: String) -> Unit
    ): () -> Unit
}

@Composable
fun useShowToast(pubSub: PubSub) {
    DisposableEffect(pubSub) {
        val unsubscribe = pubSub.subscribe(::toastStub)
        onDispose(unsubscribe)
    }
}

File: app/src/main/java/com/github/pokatomnik/frontendconf2023/examples/ex09disposableeffects/ex09DisposableEffects.ts
const toastStub = () => {
  // Показываем здесь сообщение
};

interface PubSub {
  subscribe(
    subscriber: (value: String) => void
  ): () => void;
}

const useShowToast = (props: {
  readonly pubSub: PubSub
}) => {
  const { pubSub } = props;
  React.useEffect(() => {
    const unsubscribe = pubSub.subscribe(toastStub);
    return unsubscribe;
  }, [pubSub]);
};
